doc: add style guide (#6070)
spacemeshos/go-spacemesh: STYLE.md

formatting—gofmt handles that for us.

The goal of this guide is to manage this complexity by describing in detail the
Dos and Don'ts of writing Go code at Uber. These rules exist to keep the code
base manageable while still allowing engineers to use Go language features
productively.

This guide was originally created by [Prashant Varanasi](https://github.com/prashantv) and [Simon Newton](https://github.com/nomis52) as
a way to bring some colleagues up to speed with using Go. Over the years it has
been amended based on feedback from others.

This documents idiomatic conventions in Go code that we follow at Uber. A lot
of these are general guidelines for Go, while others extend upon external
resources:

1. [Effective Go](https://go.dev/doc/effective_go)
2. [Go Common Mistakes](https://go.dev/wiki/CommonMistakes)
3. [Go Code Review Comments](https://go.dev/wiki/CodeReviewComments)

We aim for the code samples to be accurate for the two most recent minor versions
of Go [releases](https://go.dev/doc/devel/release).

All code should be error-free when run through `golint` and `go vet`. We
recommend setting up your editor to:

- Run `goimports` on save

> - Run `golint` and `go vet` to check for errors

Reviewer: We use more modern tools 🙂 

```suggestion
- Run `gofumpt` on save
- Run `make lint` or `golangci-lint` to check for errors
```

You can find information in editor support for Go tools here:
https://go.dev/wiki/IDEsAndTextEditorPlugins

## Guidelines

### Pointers to Interfaces

You almost never need a pointer to an interface. You should be passing
interfaces as values—the underlying data can still be a pointer.

An interface is two fields:

1. A pointer to some type-specific information. You can think of this as
   "type."
2. Data pointer. If the data stored is a pointer, it’s stored directly. If
   the data stored is a value, then a pointer to the value is stored.

If you want interface methods to modify the underlying data, you must use a
pointer.

### Verify Interface Compliance

Verify interface compliance at compile time where appropriate. This includes:

- Exported types that are required to implement specific interfaces as part of
  their API contract
- Exported or unexported types that are part of a collection of types
  implementing the same interface

> - Other cases where violating an interface would break users

Reviewer: I feel this should be more nuanced - yes if we implement an interface form the standard lib like `io.Writer` and `http.Handler` this makes sense. Those are interfaces that we have no control over and if we want to adhere to them a type check can be helpful to do so. However we often define our own interfaces, usually not even exporting them since they only exist for the purpose of decoupling / mocking. Generally these should be defined on the user side for a few reasons:

- prevents tight coupling between packages holding provider and user of implementation
- if interfaces / implementation is accidentally changed we still notice it right away on the user side
- if interfaces / implementation is intentionally changed its the same amount of effort either way - we generally only have one provider and few users - mocks can be regenerated

Author: I feel that having the unexported interfaces is a bit skewed in the first place. 

Ideally those interfaces should be exposed already publicly on the dependency package.

For example, if the p2p server exposes some pubsub methods through an interface, it should also generate the p2p mocks and we should consume those everywhere in the codebase where this dependency is used (instead of duplicating those interfaces endlessly, and generating redundant mocks). It generates less points of friction for testing and for handling the dependency injection.

In some cases indeed subsets of interfaces are needed but I'm not sure why not to just compose the interfaces together at the dependency level, this is not unheard of.

Another point for public interfaces is allowing us to properly do black-box testing which I feel is way superior to white-box. Ideally, white-box should be resorted to only in exceptional cases, but otherwise tests shouldn't leak package internals. Currently most of the tests don't have namespace separation and I feel it causes a lot of mess when asserting state. A lot of the necessary helpers can be put in an `export_test.go` [file](https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd) (which is a bit hacky in the first place, but at least puts a lid on hackiness).

Author: And back to the conversation about interface compliance - it would be nice to have, but we could also change it to compliance with external interfaces only for now, it is fine by me.

Reviewer: The implementation check `var _ MyInterface = MyType{}` doesn't add anything to our codebase, also without it we quickly notice that a type that should implement an interface doesn't any more because at the point of its use it will cause a compilation error.

We are not providing a library to be used by others; the interfaces we use exist for decoupling and mocking. The reason we define interfaces at the user side is because the user decides which subset of the service they want to use and/or mock.

If the implementation changes its signature we have to update all uses anyway, doesn't matter if the interface is on the user or the provider side. I don't mind trying more aggressively to go to black box tests, but even then I see no need to define interfaces at the implementation side - we want to avoid dependencies between packages. Ideally only `node` has a dependency on (almost) everything while all other packages have dependencies on (almost) nothing.

We are right now busy with refactorings because interfaces were defined on the implementation side: In the `malfeasance` package used to be (and still partially is) the interface and types for all packages creating malfeasance proofs, but those proofs contain types of those other packages:

- `activation` creates a `malfeasance.MalfeasanceProof`, inside there is a `activation.ActivationTx`
- we now have a cyclic dependency between `activation` and `malfeasance`
- the "solution" was to create more packages: `malfeasance` depends on `malfeasance/wire`, `activation/wire` and `activation` depends on `malfeasance`, `malfeasance/wire` and `activation/wire`
- its now 4 packages instead of 2 just to get around the issue of that activation depends on malfeasance for publishing proofs and malfeasance depends on activation to validate them

Author: Ok point made :) I will amend

  defer m.Unlock()

  return m.data[k]
}
```

</td><td>

```go
type SMap struct {
  mu sync.Mutex

  data map[string]string
}

func NewSMap() *SMap {
  return &SMap{
    data: make(map[string]string),
  }
}

func (m *SMap) Get(k string) string {
  m.mu.Lock()
  defer m.mu.Unlock()

  return m.data[k]
}
```

</td></tr>

<tr><td>

The `Mutex` field, and the `Lock` and `Unlock` methods are unintentionally part
of the exported API of `SMap`.

</td><td>

The mutex and its methods are implementation details of `SMap` hidden from its
callers.

</td></tr>
</tbody></table>

### Copy Slices and Maps at Boundaries

Slices and maps contain pointers to the underlying data so be wary of scenarios
when they need to be copied.

#### Receiving Slices and Maps

Keep in mind that users can modify a map or slice you received as an argument
if you store a reference to it.

<table>
<thead><tr><th>Bad</th> <th>Good</th></tr></thead>
<tbody>
<tr>
<td>

```go
func (d *Driver) SetTrips(trips []Trip) {
  d.trips = trips
}

trips := ...
d1.SetTrips(trips)

// Did you mean to modify d1.trips?
trips[0] = ...
```

</td>
<td>

```go
func (d *Driver) SetTrips(trips []Trip) {
  d.trips = make([]Trip, len(trips))

>   copy(d.trips, trips)

Reviewer 2: There's [slices.Clone](https://pkg.go.dev/slices#Clone) for that now

}

trips := ...
d1.SetTrips(trips)

// We can now modify trips[0] without affecting d1.trips.
trips[0] = ...
```

</td>
</tr>

</tbody>
</table>

#### Returning Slices and Maps

Similarly, be wary of user modifications to maps or slices exposing internal
state.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
type Stats struct {
  mu sync.Mutex
  counters map[string]int
}

// Snapshot returns the current stats.
func (s *Stats) Snapshot() map[string]int {
  s.mu.Lock()
  defer s.mu.Unlock()

  return s.counters
}

// snapshot is no longer protected by the mutex, so any
// access to the snapshot is subject to data races.
snapshot := stats.Snapshot()
```

</td><td>

```go
type Stats struct {
  mu sync.Mutex
  counters map[string]int
}

func (s *Stats) Snapshot() map[string]int {
  s.mu.Lock()
  defer s.mu.Unlock()

  result := make(map[string]int, len(s.counters))
  for k, v := range s.counters {
    result[k] = v
  }

>   return result

Reviewer: ```suggestion
  return slices.Clone(s.counters))
```

Author: you mean `maps.Clone`?

This maintains the original error type and message.
This is well suited for cases when the underlying error message
has sufficient information to track down where it came from.

Otherwise, add context to the error message where possible
so that instead of a vague error such as "connection refused",
you get more useful errors such as "call service foo: connection refused".

Use `fmt.Errorf` to add context to your errors,
picking between the `%w` or `%v` verbs
based on whether the caller should be able to
match and extract the underlying cause.

- Use `%w` if the caller should have access to the underlying error.
  This is a good default for most wrapped errors,
  but be aware that callers may begin to rely on this behavior.
  So for cases where the wrapped error is a known `var` or type,
  document and test it as part of your function's contract.
- Use `%v` to obfuscate the underlying error.
  Callers will be unable to match it,
  but you can switch to `%w` in the future if needed.

When adding context to returned errors, keep the context succinct by avoiding
phrases like "failed to", which state the obvious and pile up as the error

> percolates up through the stack:

Reviewer: NIT: i.e. errors are not log messages! 🙂 👍

  // Failure to write metrics should not
  // break the application.
  log.Printf("Could not emit metrics: %v", err)
}

```

</td></tr>
<tr><td>

**Good**: Match the error and degrade gracefully

If the callee defines a specific error in its contract,
and the failure is recoverable,
match on that error case and degrade gracefully.
For all other cases, wrap the error and return it.

Callers further up the stack will handle other errors.

</td><td>

```go
tz, err := getUserTimeZone(id)
if err != nil {
  if errors.Is(err, ErrUserNotFound) {
    // User doesn't exist. Use UTC.
    tz = time.UTC
  } else {
    return fmt.Errorf("get user %q: %w", id, err)
  }
}
```

</td></tr>
</tbody></table>

### Handle Type Assertion Failures

The single return value form of a [type assertion](https://go.dev/ref/spec#Type_assertions) will panic on an incorrect
type. Therefore, always use the "comma ok" idiom.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
t := i.(string)
```

</td><td>

```go
t, ok := i.(string)
if !ok {
  // handle the error gracefully
}
```

</td></tr>
</tbody></table>

<!-- TODO: There are a few situations where the single assignment form is

> fine. -->

Reviewer: Generally if you know the type assertion cannot fail:

- the type is the same: e.g. `type LayerID uint32` cast to `uint32` will aways work
- an `any` value is always of a certain type (and `any` is only used to satisfy some interface)
- etc.

Author: Added some comment about this. :+1:

Code running in production must avoid panics. Panics are a major source of
[cascading failures](https://en.wikipedia.org/wiki/Cascading_failure). If an error occurs, the function must return an error and
allow the caller to decide how to handle it.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
func run(args []string) {
  if len(args) == 0 {
    panic("an argument is required")
  }
  // ...
}

func main() {
  run(os.Args[1:])
}
```

</td><td>

```go
func run(args []string) error {
  if len(args) == 0 {
    return errors.New("an argument is required")
  }
  // ...
  return nil
}

func main() {
  if err := run(os.Args[1:]); err != nil {
    fmt.Fprintln(os.Stderr, err)
    os.Exit(1)
  }
}
```

</td></tr>
</tbody></table>

Panic/recover is not an error handling strategy. A program must panic only when
something irrecoverable happens such as a nil dereference. An exception to this is
program initialization: bad things at program startup that should abort the
program may cause panic.

```go
var _statusTemplate = template.Must(template.New("name").Parse("_statusHTML"))
```

Even in tests, prefer `t.Fatal` or `t.FailNow` over panics to ensure that the

> test is marked as failed.

Reviewer: Or the equivalent `assert.Fail`/`assert.FailNow` and `require.Fail`/`require.FailNow`

Author: I would personally stick to the `t.Fatal`/`t.FailNow`. I'm not sure what value do these external packages provide to such a basic operation.

Reviewer: We are already using them in tests to a high degree, the testify versions call `t.Fatal`/`t.FailNow` but add more context to the error message and print them in a visually appealing way 🙂

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
// func TestFoo(t *testing.T)

f, err := os.CreateTemp("", "test")
if err != nil {
  panic("failed to set up test")
}
```

</td><td>

```go
// func TestFoo(t *testing.T)

f, err := os.CreateTemp("", "test")

> if err != nil {
>   t.Fatal("failed to set up test")
> }

Reviewer: Since we are using testify, I'd prefer:

```suggestion
require.NoError(t, err, "failed to set up test")
```

Author: Can you please elaborate on why this is needed? I am at the opinion that this deviates from the general style-guide and personally makes things harder to read. Having this sort of deviation from the idiomatic go error handling blocks encourages all sorts of schtiks that reduce code readability in favor of developer creativity and indirection, etc. I don't think this is necessary and that we should start being lazy about writing explicit error handling blocks.

Reviewer: Lets use a different example:

```go
require.ErrorIs(t, err, io.EOF)
```

this is much more succinct than:

```go
if err == nil {
    t.Fatal("expected error")
}
if !errors.Is(err, io.EOF) {
   t.Fatal("didn't receive expected error; got %T", err)
}
```

besides this doesn't print all the information `require.ErrorIs` would print.

  cleanup tasks.

#### Exit Once

If possible, prefer to call `os.Exit` or `log.Fatal` **at most once** in your
`main()`. If there are multiple error scenarios that halt program execution,
put that logic under a separate function and return errors from it.

This has the effect of shortening your `main()` function and putting all key
business logic into a separate, testable function.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
package main

func main() {
  args := os.Args[1:]
  if len(args) != 1 {
    log.Fatal("missing file")
  }
  name := args[0]

  f, err := os.Open(name)
  if err != nil {
    log.Fatal(err)
  }
  defer f.Close()

  // If we call log.Fatal after this line,
  // f.Close will not be called.

  b, err := io.ReadAll(f)
  if err != nil {
    log.Fatal(err)
  }

  // ...
}
```

</td><td>

```go
package main

func main() {
  if err := run(); err != nil {
    log.Fatal(err)
  }
}

func run() error {
  args := os.Args[1:]
  if len(args) != 1 {
    return errors.New("missing file")
  }
  name := args[0]

  f, err := os.Open(name)
  if err != nil {
    return err
  }

>   defer f.Close()

Reviewer 2: Note that this swallows error from `f.Close()` which is not good.
There's a pattern that helps with error handling in `defer`:  https://wstrm.dev/posts/errors-join-heart-defer/
Discussion on `/r/golang`: https://www.reddit.com/r/golang/comments/11luvxm/errorsjoin_defer/
Also: https://stackoverflow.com/a/76287159
IMO it has a downside, namely that you must not forget to use named return value (which by itself is sometimes frowned upon when you don't specify the actual return values in `return`), but still it's probably better than alternatives

```go
func run() (err error) {
  args := os.Args[1:]
  if len(args) != 1 {
    return errors.New("missing file")
  }
  name := args[0]

  f, err := os.Open(name)
  if err != nil {
    return err
  }
  defer func () {
    err = errors.Join(err, f.Close())
  }()

  b, err := io.ReadAll(f)
  if err != nil {
    return err
  }

  // ...
}
```

If we're to use this pattern, it probably belongs to the section on `defer`

Reviewer: I'm not sure if I like this pattern. It doesn't handle the error. If closing fails this should in my opinion be handled where the closing happened and not further up the call stack. This pattern only achieves that the error isn't lost.

Also this introduces possible false positives:

```go
func run() (err error) {
	f, _ := os.Create("test.txt")
	defer func() {
		err = errors.Join(err, f.Close())
	}()

	f.Write([]byte("test"))
	if err := f.Close(); err != nil { // closed explicitly to handle error
		// handle err instead of just logging it
		log.Fatal(err)
	}
	return nil
}

func main() {
	fmt.Println(run())
}
```

This prints the error "file closed already" which isn't an error and should be filtered, making the `defer`red function unnecessarily complex 😕

Reviewer 2: The problem is that if we have many clauses like this in `run()`
```go
  b, err := io.ReadAll(f)
  if err != nil {
    return err
  }

  if err := doSmthWithBuffer(b); err != nil {
    return fmt.Errorf("smth: %w", err)
  }

  if err := doSmthElse(b); err != nil {
    return fmt.Errorf("smth else: %w", err)
  }
```
we would have to rewrite them like
```go
  b, err := io.ReadAll(f)
  if err != nil {
    return errors.Join(err, f.Close())
  }

  if err := doSmthWithBuffer(b); err != nil {
    return errors.Join(fmt.Errorf("smth: %w", err), f.Close())
  }

  if err := doSmthElse(b); err != nil {
    return errors.Join(fmt.Errorf("smth else: %w", err), f.Close())
  }
```
which quickly becomes error-prone, as it's easy to shoot yourself in the foot by missing `f.Close()` in one of the branches as the code evolves. Moreover, if instead of just closing the file we had some more advanced cleanup that must run even if the code panics somewhere, it will fail to run.
Basically that's the situation for which `defer` exists, but we also need to handle errors that happen during the cleanup. Good if we can just log them and that's it, but such an approach is not always applicable.

Of course care must be taken not to close the file twice and not to handle the error twice (log and return), so in some cases the `defer` approach is not appropriate. But if you need to always close the file at the end of the function, this works quite well as you don't miss any errors and still have benefits of `defer`.

EDIT: probably `io.ReadAll()` is not the best example to illustrate my point, as in most cases you can close the file immediately following the call, but let's assume `f` still needs to be open after that point

Reviewer: But you are not handling the error, you are just appending it to another - it might even be caused by the first error to begin with?

Instead of an anonymous deferred function the closing should be handled properly - if it even makes a difference to begin with 🤷 

```go
func closeWithErrorHandling(f io.Closer) {
    err := f.close()
    if err == nil {
        return
    }
    // handle error
}

func run() error {
    f, _ := os.Create("test.txt")
    defer closeWithErrorHandling(f)

    // proceed doing with `f` what needs to be done
}

Reviewer 2: This kind of error handling will not cause `run()` to return an error, and oftentimes we might want to indicate that `run` has failed to the caller instead of (or in addition to) doing any special handling in the `run`.
Of course if `run` can do all the handling we need and then return `nil` error such as log the error this pattern is unnecessary, but that is not what we always need.

So, just to be clear, the circumstances that warrant the use of the pattern are:
* a function needs to perform some cleanup when returning
* the cleanup may fail
* there are multiple `return` points in the function upon which the cleanup must be performed
* the function cannot perform all the necessary error handling on its own and must return an error if cleanup fails

This may sound a bit too specific but it's actually rather common situation

Reviewer: but `run` will return an error if it's a different one from the one caused by `Close`? The error from close doesn't need to be returned it has already been handled...

EDIT: also if another error occurs the error from the deferred function isn't helpful in the investigation of what went wrong usually. So the error only needs to be handled if everything else went well.

I think this need to handle an error in a deferred function rarely if ever comes up if the cleanup is called explicitly on the good path and will usually just lead to polluting the actual issue with noise.

Reviewer 2: Whether returning errors in `defer` in case of failure is counterproductive is debatable, depending on the situation at hand.
E.g. if we're closing the database after a failed query, normally closing the database will succeed if the problem is a simple SQL error, but if the problem has happened due to some severe database corruption, failed `Close` may give some additional hints. IoW: in the majority of cases, the cleanup should succeed even on the "unhappy" path, and in rare cases where it fails the cleanup failure may provide additional hints.

Besides, a function may have several happy paths with early returns in addition to error returns, and upon such "happy" returns we may want to perform the cleanup and let the caller know if smth went wrong at the cleanup phase, too.

Author: I'd like to chip in here because this showed up in your PR @ivan4th and I also commented on this.

I am at the opinion that if we have to add this sort of cleanup code that has all sorts of complicated error instrumentation that is just difficult to understand and is not so informative (how do you know which line in the error corresponds to which `errors.Join`, etc), it is probably cleaner and nicer to do the cleanup _once_ on the calling context (i.e. the context that called the function that has to do the convoluted cleanup logic). Having these sort of anonymous functions is really difficult to read (to whoever didn't write them usually).

As for `errors.Join`, I find it useful in a case of a task that is parallelized (like in an `errgroup`, but then again `errgroup` should do this for you anyway a-la-carte).

On a personal note: I've seen maybe a case or two  where this sort of cleanup logic was needed so far (not in this project) and it was in a p2p protocol, where developer error was very easy to do and have caused certain cases where a `return` statement could miss satisfying all sorts of other calls that had to be done (cleanup of the libp2p stream, logging, metrics, etc), which is why it was later decided to have an anonymous function that catches whether an error was caught, causing the cleanup to be made more nicely (however, the code-smell is that the protocol was already way to convoluted, and this is why this was needed in the first place).

My suggestion here is that maybe we could also do a pair/mob programming session on the code in question (in the sql vaacum PR) so that we could see together whether the `errors.Join` is needed or not.

Reviewer 2: I'm not insisting on the use of said pattern, if others don't like it, I will not use it, and I will update vacuum PR accordingly :)
If smth like that is to be used, simple `errors.Join` is probably indeed not the best approach, a cleanup errors need to be augmented, too, to make it cleaner what really happened. 

Moving cleanup to the calling context may often make sense, but in some cases it may also cause proliferation of funcs like `doSomethingInner`, `doInnerSomethingOfAnotherInnerSomething`, etc, if the only purpose of extracting a part of function is making cleanup easier. 

But in any case, simple `defer f.Close()` in the original example can be said to constitute a bug, as it may swallow a real I/O error which happens when the file is closed.

Reviewer 2: Re pair/mob programming session, that might be a good idea, too

Reviewer: > But in any case, simple defer f.Close() in the original example can be said to constitute a bug, as it may swallow a real I/O error which happens when the file is closed.

Yes this is true but if that happens, it doesn't really matter that `f.Cose()` fails:

```go
f, err := os.Open("path/to/file")
if err != nil {
    return err
}
defer f.Close()
buf, err := io.ReadAll(f)
if err != nil {
    return fmt.Error("read file: %w", err) // if close fails it doesn't really matter; read already failed, close will fail for the same cause if it does
}
// do something with `buf`
// happy case reached we now close explicitly
if err := f.Close(); err != nil {
    return fmt.Errorf("closing file: %w", err)
}
return nil
```

Reviewer 2: @fasmat my point is, the original code doesn't have any handling for `f.Close()` errors at all:
```go

  f, err := os.Open(name)
  if err != nil {
    log.Fatal(err)
  }
  defer f.Close()

  // If we call log.Fatal after this line,
  // f.Close will not be called.

  b, err := io.ReadAll(f)
  if err != nil {
    log.Fatal(err)
  }
```
Granted, we're reading the file not writing it, which kind of makes implications less severe, but still completely ignore `Close` errors is not correct. In this case, the `Close` error will be ignored even if `io.ReadAll` succeeds.

Author: So I think that @fasmat is arguing here that handling the `Close` doesn't really matter because probably the error is so severe that the application will exit anyway. From my personal experience, having a `db.Close` on some part of a component that is going to raise and exception anyway is _best-effort_ at best. If you can't `Close` a db it usually means that:
- you have some very severe I/O errors on the hardware level
- disk is out of space and you can't flush stuff to it anymore

In our problem domain, I believe that other issues are probable but I'm more familiar with the ones above. Mostly, in those cases you assume that the node is no longer able to remain consistent (once you're out of space you're screwed because you don't know whether a "large" write is successful vs "small" write which largely leaves you with the assumption of no-longer-consistent).

In those cases node-operator is needed to recover the node and usually you need to start over. Hence I support that `Close` is a best-effort and the error can be ignored. 

https://stackoverflow.com/questions/37977714/can-i-get-an-error-when-closing-a-file

  for {
    flush()
    time.Sleep(delay)
  }
}()
```

</td><td>

```go
var (
  stop = make(chan struct{}) // tells the goroutine to stop
  done = make(chan struct{}) // tells us that the goroutine exited
)
go func() {
  defer close(done)

  ticker := time.NewTicker(delay)
  defer ticker.Stop()
  for {
    select {
    case <-ticker.C:
      flush()
    case <-stop:
      return
    }
  }
}()

// Elsewhere...
close(stop)  // signal the goroutine to stop
<-done       // and wait for it to exit
```

</td></tr>
<tr><td>

There's no way to stop this goroutine.
This will run until the application exits.

</td><td>

This goroutine can be stopped with `close(stop)`,
and we can wait for it to exit with `<-done`.

</td></tr>
</tbody></table>

#### Wait for goroutines to exit

Given a goroutine spawned by the system,
there must be a way to wait for the goroutine to exit.
There are two popular ways to do this:

- Use a `sync.WaitGroup`.
  Do this if there are multiple goroutines that you want to wait for

  ```go
  var wg sync.WaitGroup
  for i := 0; i < N; i++ {
    wg.Add(1)
    go func() {
      defer wg.Done()
      // ...
    }()
  }

  // To wait for all to finish:
  wg.Wait()

>   ```

Reviewer: Or better: an `errgroup.Group`

</td></tr>
</tbody></table>

#### Specifying Slice Capacity

Where possible, provide capacity hints when initializing slices with `make()`,
particularly when appending.

```go
make([]T, length, capacity)
```

Unlike maps, slice capacity is not a hint: the compiler will allocate enough
memory for the capacity of the slice as provided to `make()`, which means that
subsequent `append()` operations will incur zero allocations (until the length
of the slice matches the capacity, after which any appends will require a resize
to hold additional elements).

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
for n := 0; n < b.N; n++ {
  data := make([]int, 0)
  for k := 0; k < size; k++{
    data = append(data, k)
  }
}
```

</td><td>

```go
for n := 0; n < b.N; n++ {
  data := make([]int, 0, size)
  for k := 0; k < size; k++{
    data = append(data, k)
  }
}
```

</td></tr>
<tr><td>

```plain
BenchmarkBad-4    100000000    2.48s
```

</td><td>

```plain
BenchmarkGood-4   100000000    0.21s
```

</td></tr>
</tbody></table>

## Style

### Avoid overly long lines

Avoid lines of code that require readers to scroll horizontally
or turn their heads too much.

> We recommend a soft line length limit of **99 characters**.

Reviewer: ```suggestion
We recommend a soft line length limit of **120 characters**.
```

That's what our current linter enforces (must not be longer than that) and we should maybe reconsider increasing that limit. We are not in the 80 and still use terminals 😉. Even if I have 2 files open at the same time next to each other on a modern wide screen display there is still unused whitespace...

Author: Yeah I totally agree. I am also on `neovim` with a 13 something inch screen and I find the 120 character to be extremely limiting and makes a lot of method calls to be very strange to read with all the line breaks. Not sure why this is needed to be so small of a value.

but it is not a hard limit.
Code is allowed to exceed this limit.

### Be Consistent

Some of the guidelines outlined in this document can be evaluated objectively;
others are situational, contextual, or subjective.

Above all else, **be consistent**.

Consistent code is easier to maintain, is easier to rationalize, requires less
cognitive overhead, and is easier to migrate or update as new conventions emerge
or classes of bugs are fixed.

Conversely, having multiple disparate or conflicting styles within a single
codebase causes maintenance overhead, uncertainty, and cognitive dissonance,
all of which can directly contribute to lower velocity, painful code reviews,
and bugs.

When applying these guidelines to a codebase, it is recommended that changes
are made at a package (or larger) level: application at a sub-package level
violates the above concern by introducing multiple styles into the same code.

### Group Similar Declarations

Go supports grouping similar declarations.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
import "a"
import "b"
```

</td><td>

```go
import (
  "a"
  "b"
)

> ```

Reviewer: For imports this is enforced by `golangci-lint` and can be autoformated with `gci` (or `golangci-lint --fix`) The order is:

```go
import (
    "io"
    "math" // stdlibrary

    "github.com/some/library"
    "github.com/stretchr/testify" // packages from other modules

    "github.com/spacemeshos/go-spacemesh/activation" // packages from same module
    "github.com/spacemeshos/go-spacemesh/proposal"
)
```

This is also enforced by `make lint`.

Author: So should we remove the entry for this? (in case it is already enforced by static analysis)

</td></tr>
</tbody></table>

This also applies to constants, variables, and type declarations.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go

const a = 1
const b = 2

var a = 1
var b = 2

type Area float64
type Volume float64
```

</td><td>

```go
const (
  a = 1
  b = 2
)

var (
  a = 1
  b = 2
)

type (
  Area float64
  Volume float64
)

> ```

Reviewer: This is actually enforced by `make lint`

Author: Should remove?

Groups are not limited in where they can be used. For example, you can use them
inside of functions.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
func f() string {
  red := color.New(0xff0000)
  green := color.New(0x00ff00)
  blue := color.New(0x0000ff)

  // ...
}
```

</td><td>

```go
func f() string {
  var (
    red   = color.New(0xff0000)
    green = color.New(0x00ff00)
    blue  = color.New(0x0000ff)
  )

  // ...
}
```

</td></tr>
</tbody></table>

Exception: Variable declarations, particularly inside functions, should be
grouped together if declared adjacent to other variables. Do this for variables
declared together even if they are unrelated.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
func (c *client) request() {
  caller := c.name
  format := "json"
  timeout := 5*time.Second
  var err error

  // ...
}
```

</td><td>

```go
func (c *client) request() {
  var (
    caller  = c.name
    format  = "json"
    timeout = 5*time.Second
    err error
  )

  // ...
}
```

</td></tr>
</tbody></table>

### Import Group Ordering

There should be two import groups:

- Standard library
- Everything else

> This is the grouping applied by goimports by default.

Reviewer: See comment above

Author: ditto

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
import (
  "fmt"
  "os"
  "go.uber.org/atomic"
  "golang.org/x/sync/errgroup"
)
```

</td><td>

```go
import (
  "fmt"
  "os"

  "go.uber.org/atomic"
  "golang.org/x/sync/errgroup"
)
```

</td></tr>
</tbody></table>

### Package Names

When naming packages, choose a name that is:

- All lower-case. No capitals or underscores.
- Does not need to be renamed using named imports at most call sites.
- Short and succinct. Remember that the name is identified in full at every call
  site.
- Not plural. For example, `net/url`, not `net/urls`.

> - Not "common", "util", "shared", or "lib". These are bad, uninformative names.

Reviewer: ❤️

try to reduce the scope.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
if data, err := os.ReadFile(name); err == nil {
  err = cfg.Decode(data)
  if err != nil {
    return err
  }

  fmt.Println(cfg)
  return nil
} else {
  return err
}
```

</td><td>

```go
data, err := os.ReadFile(name)
if err != nil {
   return err
}

if err := cfg.Decode(data); err != nil {
  return err
}

fmt.Println(cfg)
return nil
```

</td></tr>
</tbody></table>

### Avoid Naked Parameters

Naked parameters in function calls can hurt readability. Add C-style comments
(`/* ... */`) for parameter names when their meaning is not obvious.

<table>
<thead><tr><th>Bad</th><th>Good</th></tr></thead>
<tbody>
<tr><td>

```go
// func printInfo(name string, isLocal, done bool)

printInfo("foo", true, true)
```

</td><td>

```go
// func printInfo(name string, isLocal, done bool)

printInfo("foo", true /* isLocal */, true /* done */)
```

</td></tr>
</tbody></table>

Better yet, replace naked `bool` types with custom types for more readable and
type-safe code. This allows more than just two states (true/false) for that

> parameter in the future.

Reviewer: Even better: except for methods that only take a single bool as parameter, booleans generally indicate that a method/function should be two instead.

Author: Can you give an example just for the sake of clarity?

Reviewer: ```go
func CalcRank(wins int, losses int, isTournament bool) {
    if(isTournament) {
        return wins / (wins + losses)
    }
    
    return (wins) / (losses)
}
```

should be 2 functions: `CalcRank` and `CalcTournamentRank`. There are (rare) exceptions, but usually a boolean indicates that it should be two methods/functions instead.

  },
  {
    give:     "192.0.2.0:http",
    wantHost: "192.0.2.0",
    wantPort: "http",
  },
  {
    give:     ":8000",
    wantHost: "",
    wantPort: "8000",
  },
  {
    give:     "1:8",
    wantHost: "1",
    wantPort: "8",
  },
}

for _, tt := range tests {
  t.Run(tt.give, func(t *testing.T) {
    host, port, err := net.SplitHostPort(tt.give)
    require.NoError(t, err)
    assert.Equal(t, tt.wantHost, host)
    assert.Equal(t, tt.wantPort, port)
  })
}
```

</td></tr>
</tbody></table>

Test tables make it easier to add context to error messages, reduce duplicate
logic, and add new test cases.

We follow the convention that the slice of structs is referred to as `tests`
and each test case `tt`. Further, we encourage explicating the input and output
values for each test case with `give` and `want` prefixes.

```go
tests := []struct{
  give     string
  wantHost string
  wantPort string
}{
  // ...
}

for _, tt := range tests {
  // ...
}
```

#### Avoid Unnecessary Complexity in Table Tests

Table tests can be difficult to read and maintain if the subtests contain conditional
assertions or other branching logic. Table tests should **NOT** be used whenever

> there needs to be complex or conditional logic inside subtests (i.e. complex logic inside the `for` loop).

Reviewer: ```suggestion
there needs to be complex or conditional logic inside subtests (i.e. complex logic inside a `for` loop
or changed behaviour of subtests due to an `if`).
```

  require.NoError(t, err)
  assert.Equal(t, "want", got)
}

func TestShouldCallYAndFail(t *testing.T) {
  // setup mocks
  ctrl := gomock.NewController(t)
  xMock := xmock.NewMockX(ctrl)

  yMock := ymock.NewMockY(ctrl)
  yMock.EXPECT().Call().Return("YResponse", nil)

  _, err := DoComplexThing("inputY", xMock, yMock)
  assert.EqualError(t, err, "Y failed")
}
```
</td></tr>
</tbody></table>

This complexity makes it more difficult to change, understand, and prove the
correctness of the test.

While there are no strict guidelines, readability and maintainability should
always be top-of-mind when deciding between Table Tests versus separate tests
for multiple inputs/outputs to a system.

#### Parallel Tests

Parallel tests, like some specialized loops (for example, those that spawn
goroutines or capture references as part of the loop body),
must take care to explicitly assign loop variables within the loop's scope to
ensure that they hold the expected values.

```go
tests := []struct{
  give string
  // ...
}{
  // ...
}

for _, tt := range tests {
  tt := tt // for t.Parallel
  t.Run(tt.give, func(t *testing.T) {
    t.Parallel()
    // ...
  })
}

> ```

Reviewer: This has become obsolete with `Go 1.22` generally loop variables don't need to be copied any more even when using parallel execution.

struct.

```go
type options struct {
  cache  bool
  logger *zap.Logger
}

type Option interface {
  apply(*options)
}

type cacheOption bool

func (c cacheOption) apply(opts *options) {
  opts.cache = bool(c)
}

func WithCache(c bool) Option {
  return cacheOption(c)
}

type loggerOption struct {
  Log *zap.Logger
}

func (l loggerOption) apply(opts *options) {
  opts.logger = l.Log
}

func WithLogger(log *zap.Logger) Option {
  return loggerOption{Log: log}
}

// Open creates a connection.
func Open(
  addr string,
  opts ...Option,
) (*Connection, error) {
  options := options{
    cache:  defaultCache,
    logger: zap.NewNop(),
  }

  for _, o := range opts {
    o.apply(&options)
  }

  // ...
}
```

Note that there's a method of implementing this pattern with closures but we
believe that the pattern above provides more flexibility for authors and is
easier to debug and test for users. In particular, it allows options to be
compared against each other in tests and mocks, versus closures where this is
impossible. Further, it lets options implement other interfaces, including
`fmt.Stringer` which allows for user-readable string representations of the

> options.

Reviewer 2: As it was mentioned by @acud elsewhere, overusing options as we do it now may not always be good, so maybe some discussion needed regarding the alternatives, and some examples showing when to use options and when to use a simple approach.
Another problem is that the approach to options which is described here differs from what's in use in the most of go-spacemesh codebase at this moment, which is using closure-based options. Although the advantages described in this paragraph may be valid, they may not be always worth the cost of adding verbosity to already verbose options mechanism (which is essentially a verbose kwargs replacement), and also in this case we will either have inconsistent option mechanisms in the codebase, or we will have to refactor a lot of existing code.

Reviewer: The difference between "closure options" and "functional options" is only that the closure options manipulate the created object directly, while the functional operate on an intermediary object that only holds configuration and than use that to construct the actual service. Sure in the details it is different but from the point of using them I see little difference.

Functional options should be used for _optional_ parameters and thats where they are currently overused in our codebase. In many places the parameters that are passed to the services are not optional and not setting them will cause the instantiation of the object to fail.

I've occasionally done this as well primarily for readability:

```go
func NewService(
    db,
    lg,
    clk,
    sync,
    h,
)
```

is just much less readable than

```go
func NewService(
    WithDB(db),
    WithLogger(lg),
    WithClock(clk),
    WithSyncer(sync),
    WithHandler(h),
)
```

Sure variable naming can be improved too to mitigate this issue but functional options have an additional benefit for testing where constructing many services with slightly different configuration becomes much easier:

```go
defaultOpts := []Opts{
    WithDB(db),
    WithLogger(lg),
    WithClock(clk),
    WithSyncer(sync),
    WithHandler(h),
}

testservice1 := NewService(defaultOpts..., WithHandler(mockHandler))
testservice2 := NewService(defaultOpts..., WithSyncer(mockSyncer))
// ... etc.
```

Reviewer 2: Problem is that the paragraph I commented recommends against using closure-based options. Thus we either update the guide, refactor our code or accept it as an inconsistency

As of options being useful in many cases, I can only agree with that, but there are also arguments against using them too much:
https://github.com/spacemeshos/go-spacemesh/pull/6085#discussion_r1660225344

I for one don't find these arguments to be applicable to all cases as having 10 different constructors to avoid the use of the option pattern doesn't sound too good to me, but there probably needs to be some kind of balance

Author: So I'm with @ivan4th on this. Personally I find this sort of pattern to be very confusing because it is also allowing for irregular sequence of setting things in a constructor. It has the sort of flexibility that the language intentionally tries to avoid for various reasons.
Also having these sort of parameters has the unfortunate consequence that a lot of the components in the code-base do things in a very complex way. For example:
- component defines it's _own_ configuration with a default configuration which is then _potentially_ overridden by the instantiator (BAD) - configuration should never be an optional parameter
- logger is defined by the component but then _potentially_ overridden by the instantiator - why should a logger parameter be optional? for me this is a mandatory thing and that instantiator should know whether the logger that should be passed is a noop logger or not
- DB is not optional and should not be marked as such by using this pattern.

IMO the "internal vocalisation" of how we read code shouldn't affect the way we write it. And yes I agree that sometimes having a constructor call with a lot of arguments can become cumbersome but you at least know for sure what is needed and it usually also breaks-fast so one can fix it fast.

So TL;DR my personal opinion is that I find this pattern less readable and that it adds an unreasonable amount of mental taxation reading the code (and resolving its dependencies)

help to catch the most common issues and also establish a high bar for code
quality without being unnecessarily prescriptive:

- [errcheck](https://github.com/kisielk/errcheck) to ensure that errors are handled
- [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) to format code and manage imports
- [golint](https://github.com/golang/lint) to point out common style mistakes
- [govet](https://pkg.go.dev/cmd/vet) to analyze code for common mistakes
- [staticcheck](https://staticcheck.dev) to do various static analysis checks

### Lint Runners

We recommend [golangci-lint](https://github.com/golangci/golangci-lint) as the go-to lint runner for Go code, largely due
to its performance in larger codebases and ability to configure and use many
canonical linters at once. This repo has an example [.golangci.yml](https://github.com/uber-go/guide/blob/master/.golangci.yml) config file
with recommended linters and settings.

golangci-lint has [various linters](https://golangci-lint.run/usage/linters/) available for use. The above linters are
recommended as a base set, and we encourage teams to add any additional linters

> that make sense for their projects.

Reviewer: We have a `.golangci-lint` configuration that we use for this project. Developers should use the configuration we use when linting their code for a contribution. 🙂

Author: Added as a comment :+1:
